### PR TITLE
ROX-25266: Deploy compatibility test against supported versions

### PR DESCRIPTION
**Pull request by jschnath**
### Description

Implements the logic for running non-groovy compatibility tests and extracts some common functionality between the `gke_nongroovy_compatibility_tests` target and the `gke_version_compatibility_tests` target.

Related openshift PR to add the new test target: https://github.com/openshift/release/pull/55872

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

Added a main() function to `get_compatibility_test_tuples.py` to manually confirm that its output remains correct. Next is running `gke_version_compatibility_tests` to ensure that they still function as intended. For testing the new target itself I think this PR needs to be merged first, and then I will be able to run `/pj-rehearse ...` on the openshift PR and look at the test output.

### CODE REVIEW

**Comment by vikin91** [comment]
@jschnath I am getting a lot of emails with updates on this and other PRs. I use those e-mails to form a queue of things that I want to review. I was wondering if you could maybe mark this PR as draft and use `/test all` for as long as you need to work on it? You could then undraft it when it is ready for review and that would be a ping for the team to review this.

**Comment by jschnath** [comment]
Damn, making this draft interrupted my CI runs.
/test all
/test gke-nongroovy-compatibility-tests
/test gke-version-compatibility-tests

File: .openshift-ci/get_compatibility_test_tuples.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import sys
+from collections import namedtuple
+from pathlib import Path
+
+from get_latest_helm_chart_versions import (
+    get_supported_helm_chart_versions,
+    get_latest_helm_chart_version_for_specific_release,
+)
+
+
```
Comment (jschnath) [inline .openshift-ci/get_compatibility_test_tuples.py:17]: Add a comment here explaining what this does

Comment (jschnath) [reply .openshift-ci/get_compatibility_test_tuples.py:17]: Also explain the return / error behavior

File: tests/Makefile [hunk 2]
```diff
@@ -10,6 +10,13 @@ all: test
 	@echo "+ $@"
 	@$(MAKE) report JUNIT_OUT=all-tests-results
 
+.PHONY: compatibility-tests
+compatibility-tests:
+	@echo "+ $@"
+	@GOTAGS=$(GOTAGS),test,test_compatibility $(TOPLEVEL)/scripts/go-test.sh -cover $(TESTFLAGS) -v $(shell go list -e ./... | grep -v generated | grep -v vendor) 2>&1 | tee test.log
+	@$(MAKE) report JUNIT_OUT=compatibility-tests-results
+
+
```
Comment (jschnath) [inline tests/Makefile:19]: Remove an empty line

File: tests/common.go [hunk 3]
```diff
@@ -172,6 +175,9 @@ func waitForDeployment(t testutils.T, deploymentName string) {
 				log.Errorf("Error retrieving deployments: %s", err)
 				continue
 			}
+			for i, deployment := range deployments {
+				log.Infof("Deployment %d: %s in namespace %s, id: %s", i, deployment.Name, deployment.Namespace, deployment.GetContainers()[0].GetImage().GetId())
+			}
```
Comment (jschnath) [inline tests/common.go:181]: Remove log lines

File: tests/e2e/run-compatibility.sh [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+    rm -f FAIL
+
+    prepare_for_endpoints_test
+
+    # Give some time for stackrox to be reachable
+    wait_for_api
+
+    info "API version compatibility tests"
+    if pr_has_label "ci-release-build"; then
```
Comment (jschnath) [inline tests/e2e/run-compatibility.sh:74]: Add this label to the PR and run to test this

File: tests/e2e/run-compatibility.sh [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+    junit_wrap "roxctl-k8s-context" "roxctl --use-current-k8s-context test" "" \
+        "$ROOT/tests/roxctl/roxctl-k8s-context.sh"
+
+    junit_wrap "roxctl-helm-chart-generation" "roxctl helm-chart-generation test" "" \
+        "$ROOT/tests/roxctl/helm-chart-generation.sh"
+
+    CA="$SERVICE_CA_FILE" junit_wrap "roxctl-yaml-verification" "roxctl yaml-verification test" "" \
+        "$ROOT/tests/yamls/roxctl_verification.sh"
+}
```
Comment (jschnath) [inline tests/e2e/run-compatibility.sh:148]: Remove these

File: tests/e2e/run-compatibility.sh [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+test_preamble() {
+    require_executable "roxctl"
+
+    export ROX_PLAINTEXT_ENDPOINTS="8080,grpc@8081"
+    export ROXDEPLOY_CONFIG_FILE_MAP="$ROOT/scripts/ci/endpoints/endpoints.yaml"
+    export TRUSTED_CA_FILE="$ROOT/tests/bad-ca/root.crt"
+}
+
```
Comment (jschnath) [inline tests/e2e/run-compatibility.sh:95]: Potential followup: Remove code duplication, the same function exists in run.sh for the Go e2e tests

// Or just add a TODO comment for now

File: tests/e2e/run.sh [hunk 7]
```diff
@@ -34,8 +34,8 @@ test_e2e() {
     info "Creating mocked compliance operator data for compliance v1 tests"
     "$ROOT/tests/complianceoperator/create.sh"
```
Comment (jschnath) [inline tests/e2e/run.sh:36]: If deploy_stackrox is run before deploy_optional_e2e_components sensor will have an unnecessary restart, add a comment here explaining this

**Comment by jschnath** [comment]
Addressed synchronous review feedback from @vikin91 
Re-running with ci-release-build label

/test all
/test gke-nongroovy-compatibility-tests
/test gke-version-compatibility-tests

File: tests/e2e/run-compatibility.sh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    store_test_results "tests/compatibility-tests-results" "${compatibility_dir}"
+    [[ ! -f FAIL ]] || die "compatibility tests failed for Central v${short_central_tag}, Sensor v${short_sensor_tag}"
+
+    cd "$ROOT"
+
+    collect_and_check_stackrox_logs "/tmp/compatibility-test-logs" "${compatibility_dir}/initial_tests"
+}
+
+# Duplicate function with run.sh
+# TODO: Remove duplication
```
Comment (vikin91) [inline tests/e2e/run-compatibility.sh:89]: It would be cool to open a ticket for this in housekeeping and put it here in format we use: `TODO(ROX-####)`

**Review by vikin91 — APPROVED** [review]
We did a sync review of that and I see that all my remarks are addressed. Moreover, I see that `gke-nongroovy-compatibility-tests` is passing when run with `ci-release-build` label, so I think this is good to be merged.

File: tests/e2e/run-compatibility.sh (hunk 8 above)
Comment (jschnath) [reply tests/e2e/run-compatibility.sh]: Done 👍

**Comment by jschnath** [comment]
I'll go ahead and merge this once all required tests have passed. After that I'll rebase https://github.com/stackrox/stackrox/pull/12556 to master and it'll be much easier to review.

**Pull request merged**